ranged-based `for` for user-defined types (#1885)
carbon-language/carbon-lang: proposals/p1885.md

This proposal exposes an Iterable-type interface, relying on 2 new interfaces:
`Iterable` and `Iterator`.

A user implements these interfaces to enable range-based `for` support. Carbon
would make use of the `Iterable` to retrieve an `Iterator` and iterate until
completed or interrupted.

### Iterable

The `Iterable` interface represents a generic sequential-access type of
container. It relies on two new interfaces:

-   `Iterable`: Interface for a type that can return an iterator and be iterated
    over
-   `Iterator`: Interface for an iterator-like object

**Note:** This interface makes use of `Optional` type, and assumes its future
definition. The usage highlighted here is hypothetical.

#### Range-based `for` behavior

-   Get an iterator with `GetIterator()`
-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
    -   If the `Optional` is null, break the loop. Otherwise return the value
        held by the Optional.

#### Interface and usage

```carbon
// Basic immutable iterator
interface Iterator {
    let ElementType:! Type;
    // Needed in the broader context of Iterators and C++ interop
    fn Get[me: Self]() -> Optional(ElementType);

>     fn Next[addr me: *Self]() -> Optional(ElementType);

Reviewer: Could we define an `Advance` method that moves the iterator forward but doesn't return a value, and then give `Next` a `final` definition in terms of `Advance` and `Get`? That seems more orthogonal and easier to implement, and exposing the ability to increment without the overhead of returning a value could be a useful optimization in some cases.

Author: Agree, done

(and `for`) support for a custom interface.

#### Possible interface for 'Consumable' containers

Below is a suggestion for future 'consumable' containers. The `Consumable`
interface would represent a single-use sequential access type of container.
Items can be read (consumed) once, and are not expected to be accessible anymore
after usage.

**Note:** This interface makes use of `Optional` type, and assumes its future
definition. The usage highlighted here is hypothetical.

##### Range-based `for` behavior

-   For each iteration, call `Consume() -> Optional(ElementType)`
-   If the `Optional` is null, break the loop. Otherwise return the value held
    by the Optional.

##### Interface and usage

```carbon
// Carbon definition
interface Consumable {
    let ElementType:! Type
    fn Consume[me: Self]() -> Optional(ElementType);
}

// Usage
class MyContainer {
    ...
    impl as Consumable where .ElementType = i32 {
        fn Consume[me: Self]() -> Optional(ElementType) {
            ...
        }
    }
}

var container: MyContainer = {...};
for (var a: auto in container) {}
// `container` is now empty
```

> ### Limitations and future language improvements

Reviewer: We should probably also mention language support for `const`, and language support for user-defined l-value expressions, because those will determine how iterators should support mutability.

The design of this proposal would benefit from future support for:

> -   `Optional` (required)

Reviewer: I think we should specifically call out the performance issue of needing to copy the element, either here or as a separate bullet point, because it's a major limitation of this proposal, and an area where we're actively discussing language changes to address it.

Author: Done, and updated interface to reomve `Optional`

-   [Details](#details)
    -   [Indexable](#indexable)
        -   [Range-based `for` behavior](#range-based-for-behavior)
        -   [Interface and usage](#interface-and-usage)
        -   [Implicit `[]` support](#implicit--support)
        -   [Mutable access](#mutable-access)
    -   [Consumable](#consumable)
        -   [Range-based `for` behavior](#range-based-for-behavior-1)
        -   [Interface and usage](#interface-and-usage-1)
        -   [TODO](#todo)
    -   [Iterable](#iterable)
        -   [Range-based `for` behavior](#range-based-for-behavior-2)
        -   [Interface and usage](#interface-and-usage-2)
        -   [Iterator: `begin()` & `end()`](#iterator-begin--end)
        -   [Mutable, Bidirectional, and others](#mutable-bidirectional-and-others)
        -   [Chaining & Lazy evaluation](#chaining--lazy-evaluation)
    -   [Limitations and future language improvements](#limitations-and-future-language-improvements)
    -   [TODO](#todo-1)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Using interface type parameters](#using-interface-type-parameters)
    -   [Other](#other)

<!-- tocstop -->

> 

Reviewer 2: Proposals as of #1895 include an abstract:

```suggestion

# Abstract

The goal of this proposal is to provide a way for user-defined types to
support range-based iteration with `for`. The proposed solution exposes
3 interfaces that can be implemented by user types to enable support for
ranged-`for` loops.

```

Author: Added

## Problem

The current `for` proposal does not define a way for user-defined types to
support range-based for loops. Adding support would make it significantly easier
to define custom containers, as well as containers in Carbon standard library.

The goals for the solution includes:

-   Expressive and easy to implement
-   Minimal assumptions and requirements on the user-defined type
-   Minimal impact to maintain and use the type after (loose coupling, minimal
    impact on interface, type content, ...)
-   As well as the standard: Easy to use right, hard to use incorrectly,
    performance, don't pay for what you don't want, ...

While keeping in mind for the future:

-   Compatible with future algorithms (`sort`, `filter`, `transform`, ...)
-   Suitable for chaining (`std::ranges::views`)
-   Suitable for lazy evaluation / generators

> 

Reviewer 2: Would be good to talk about use cases here. For example, there are lots of containers that have multiple ways of iterating through them:
- forward and reverse
- maps and hash-maps have items, keys, and values

It would also be worth talking about synthetic data like iterating through the numbers 0 to 1,000,000.

Author: Done

## Background

The original proposal of `for` loops were discussed in
[p0353](/proposals/p0353.md), and the basic design documented in
[control_flow/loops#for](/docs/design/control_flow/loops.md#for).

The resulting syntax is:

> `for (` _var declaration_ `in` _expression_ `) {` _statements_ `}`

The original `for` proposal does not define a way for user-defined types to
opt-in to support range-based for loops. An approach is proposed in this
document.

> 

Reviewer 2: Maybe talk about the advantages of the cursor versus iterator approaches. Cursors are things that only make sense in the context of a container, so an integer could be a cursor marking a position in an array. Iterators are things that have all their context, like a pointer marking a position in an array, except possibly when to stop. I have heard Carbon lead @chandlerc say he favors the cursor approach because it makes it easier to have bounds checking in debug mode (and possibly other benefits).

Author: Added a small paragraph to that effect. Do you think this challenges the iterator approach in this proposal? I can imagine a situation where this could work provided the "container" implements a specific 'indexable' interface (removed from here, and a WIP by @geoffromer IIRC), that would likely put an unecessary random-access constraint on the said user-container.

Reviewer: I just realized one major advantage of the "cursor" model over the "iterator" model: it's really hard to see how array r-values could have iterators, since r-values don't have addresses, and an iterator is the functional equivalent of a pointer.

Of course, this problem goes beyond `for` loops: it implies we can't run STL algorithms over array r-values, so we might need to solve this some other way anyway. But I think the cursor model needs to at least be in "alternatives considered", if not the primary proposal.

Author: Switched to cursor, and iterator now in alternatives. Mentioned R-value simplification.

    console.log(pet); // "Cat", "Dog", "Hamster"
}
```

Using the `.iterator` property, Typescript allow to lazily generate data easily
using functions, classes, or objects:

```ts
const reverse = (arr) => ({
    [Symbol.iterator]() {
        let i = arr.length;
        return {
            next: () => ({
                value: arr[--i],
                done: i < 0,
            }),
        };
    },
});
```

#### Go

Go has non-standard support for range-based with custom types only.

https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-over-a-custom-type

## Proposal

Provide interfaces that can be implemented by user types to enable support for
ranged-for loops.

We propose three interfaces:

-   [**Indexable**](#indexable): Implements index-based random access
-   [**Iterable**](#iterable): Implements iterator-based sequential access
-   [**Consumable**](#consumable): Implements single-use sequential access

Users would benefit from providing an `external impl`, to limit class members
pollution.

## Details

This proposal covers support for 2 different type of containers:

-   Random access:
    -   `Indexable`
-   Sequential access:

>     -   `Iterable`, `Iterator`

Reviewer 2: Having multiple interfaces raises the question of what happens if you implement more than one? Since `for` loops won't ever access the elements except sequentially, it seems better to say something like:

`for` will always use the sequential interfaces. There is also a random access interface. If a type implements the random access interface, an implementation of the sequential interface will be provided automatically, using a `final` blanket `impl`.

Author: Right. @zygoloid had a similar comment, I parked `Consumable` in 00ef294, and mentioned possibility for blanket `impl`

>     -   `Consumable`

Reviewer 2: If we are to support `for` loops that consume the sequence, it seems like there should be some indication in the code that we are consuming (switching to this different interface). This is both valuable as documentation to readers of the code and clarifies what happens if both the sequential and consuming interfaces are implemented for a type.

Author: Parked Consumable in https://github.com/carbon-language/carbon-lang/commit/00ef2942718fff07346a7ceed16e424f182a98bf

Due to the semantic and functional difference between these interfaces, they

> should not conflict with the [One way](/docs/project/principles/one_way.md).

Reviewer 2: ```suggestion
should not conflict with the
 ["one way" principle](/docs/project/principles/one_way.md).
```

In absence of strong requirements from the algorithm, this would require future
algorithms to:

-   A. explicitly support multiple interfaces: expressive, and specialize
    implementations based on type features (likely preferred)

> -   or B. define adapters or implicit conversions to respectively simplify

Reviewer 2: ```suggestion
    implementations based on type features (likely preferred), or
-   B. define adapters or implicit conversions to respectively simplify
```

    algorithms implementation or end-user usage

> ### Indexable

Reviewer: Designing an indexing API for Carbon turns out to be surprisingly thorny -- see #1356. It might be better to leave this piece as future work.

Author: Thanks for linking that, I missed this proposal. Yes especially as #1356 is not approved / merged yet, this seems backwards to address it here.
Will move to "Alternative considered"

Author: `Indexable` was moved to the last section of solutions considered.

`Indexable` represents a container that supports index-based random access.

#### Range-based `for` behavior

-   Get container size (`size`)
-   Iterate from `[0;size[`, returning current item at each iteration

#### Interface and usage

> ```rust

Reviewer 2: Carbon code should use ` ```carbon ` even if that isn't supported yet.

// Carbon definition

> interface Indexable() {

Reviewer 3: ```suggestion
interface Indexable {
```
You don't need the parens here if the interface has no parameters.

    let ElementType:! Type
    fn GetSize[me: Self]() -> u32;

>     fn GetAt[me: Self](pos: u32) -> ElementType;

Reviewer: This would imply that Carbon containers that want to allow more than 2^32 elements can't support indexing. So for example, a C++ `std::vector` couldn't support indexing in Carbon. That seems like an unfortunate restriction. Also, unsigned array indices are prone to bugs, so it seems like this should be `i64`, or maybe the Carbon equivalent of `ptrdiff_t`, if we have such a thing.

In any event, it seems like it'd be a substantial regression from C++ if Carbon's `[]` syntax only supported one particular index type, so I think this type needs to be either a parameter or an associated type, rather than being fixed by the interface. But that raises the question of how to determine the start index value of an arbitrary type, and how to increment it. That's solvable with one or two more interfaces, but I'm not sure you want this proposal to get that far off into the weeds. This seems like another reason to leave this part as future work.

Author: That's a good point. The index limitation wasn't intentional when writing this proposal, more an initial placeholder that stayed. For which a parameter or associated type would be the way to go.

Based on the previous comment, will move the entire section to "Alternatives considered"

}

// Usage
class MyArray {
    ...
    impl as Indexable where .ElementType = i32 {
        fn GetSize[me: Self]() -> u32 { ... };
        fn GetAt[me: Self](pos: u32) -> ElementType { ... };
    }
}

var array: MyArray = {...};
for (var element: auto in array) {
    ...
}
```

#### Implicit `[]` support

A related added value would be the support for index-based access and assignment
with the `[]` operator, as a 1D array, for `Indexable` types. In addition to
basic single-element operations, this opens the door for future more advance
uses, as proposed in
[PR-1928](https://github.com/carbon-language/carbon-lang/pull/1928).

The type-checker would however not be able to validate `Indexable`-based array,
in comparison to fixed arrays:

```rust
fn Foo[T:! Indexable](array: T) {
    array[1] = array[2]; // May or may not exist
}
```

#### Mutable access

TODO

### Consumable

An interface representing single-use sequential access type of container. Items

> ca be read (consumed) once, and are not expected to be accessible anymore after

Reviewer 2: ```suggestion
can be read (consumed) once, and are not expected to be accessible anymore after
```

> usage.

Reviewer 3: Some examples of types that might implement this would be useful to me here. Certainly iterator types could implement this, but it seems a little unusual to treat an iterator as a range. Similarly, destructive traversal over a linked list could implement this (and a destructive traversal might in fact be better than traversal followed by destruction due to improved locality). I could also imagine this being used to good effect when implementing a generator.

Is there an example of a common vocabulary type that you'd expect to implement this?

Author: This is a bit different yes, and steps into data ownership territory. The two use cases I had in mind were generators and producer/consumer scenarios, and having an interface to make this explicit.
In terms of common types, this would apply for sure to queues and stacks, as well as any iterable (in the general sense). Generators too, though the semantic added value seems a bit more limited.

From a user standpoint, this could be something like:
```
for (e: auto in queue as Consumable) {
  ...
}
```
Which conveys the intent, independent of the container. In this example I would expect it to just `pop` until empty.
It is definitely a bit more unusual, so I'm not opposed to moving it to a different proposal, changing or dropping it.

And I will add some examples to the description here if this is deemed still valuable.

Author: Parked `Consumable` in 00ef294

**Note:** This interface makes use of `Optional` type, and assumes its future
definition. The usage highlighted here is hypothetical.

#### Range-based `for` behavior

-   For each iteration, call `Consume() -> Optional(ElementType)`
-   If the `Optional` is null, break the loop. Otherwise return the value held
    by the Optional.

#### Interface and usage

```rust
// Carbon definition

> interface Consumable() {

Reviewer 3: ````suggestion
```carbon
// Carbon definition
interface Consumable {
````

    let ElementType:! Type
    fn Consume[me: Self]() -> Optional(ElementType);
}

// Usage
class MyContainer {
    ...
    impl as Consumable where .ElementType = i32 {
        fn Consume[me: Self]() -> Optional(ElementType) {
            ...
        }
    }
}

var container: MyContainer = {...};
for (var a: auto in container) {}
// `container` is now empty
```

#### TODO

-   Size method?

### Iterable

The `Iterable` interface represents a generic sequential-access type of
container. It relies on two new interfaces:

-   `Iteratable`: Interface for a type that can return an iterator and be
    iterated over
-   `Iterator`: Interface for an iterator-like object

**Note:** This interface makes use of `Optional` type, and assumes its future
definition. The usage highlighted here is hypothetical.

#### Range-based `for` behavior

-   Get an iterator with `GetIterator()`
-   For each iteration, call `Next() -> Optional(ElementType)` on the iterator
    -   If the `Optional` is null, break the loop. Otherwise return the value
        held by the Optional.

> #### Interface and usage

Reviewer 3: It seems to me that `Iterator` is strictly more general than the other two interfaces, and that we can provide
```
impl forall [T:! Consumable] T as Iterable;
impl forall [T!: Indexable] T as Iterable;
```
in the library. That'd allow us to define the semantics of `for` entirely in terms of `Iterable`, rather than having there be three different interfaces that the library knows about.

Is that possible, or is there some subtlety that I'm missing here? If it is possible, it seems desirable to keep the language <-> library interface simpler.

Author: Absolutely, yes. I think there would be a small performance hit due to the indirection. Would that be acceptable? That case that might be a small gotcha of the language for people really performance-focused.

If it is, we might want to be resolve it by just accepting one single interface, and letting users take care of adapting for `Iterable` if necessary. I agree that the simpler the lib interface the better.

Reviewer 3: I think we should assume that the indirection won't introduce any additional runtime cost in optimizing build modes. And I think there's real value in making the language interface for `for` as simple and minimal as possible, and defining any higher-level abstractions in the library.

Author: I parked `Consumable` in 00ef294, and noted possible compatibility with `forall`.

```rust
// Basic immutable iterator

> interface Iterator() {

Reviewer 3: ````suggestion
```carbon
// Basic immutable iterator
interface Iterator {
````

    let ElementType:! Type;

>     fn Next[me: Self]() -> Optional(ElementType);

Reviewer 3: I wonder if there's a performance cost here in using `Optional(ElementType)`. Specifically, for something like `for (var x: LargeThing in sequence)`, do we pay any additional cost for a copy out  of an `Optional(T)` that we might avoid if `Next` returned a `T` directly (with a separate `HasNext` function)? In C++, if `sequence`'s iterators create new `LargeThing`  objects on the fly, that loop would perform no copies, but this iterator design seems to force a copy.

On the other hand, coupling the `Next` and `HasNext` functions seems very nice from an ergonomics and semantics perspective.

Author: It is definitely going to boil down to preference from an ergonomics and performance standpoint. If we put an empasis on performance, this might just not be an option. Offering a `SafeIterator` might do it, with the annoying multiplication of interfaces as a side effect (`SafeMutableIterator`?)

Author: Removed `Optional` and added to "Alternative considered"

}

// Iterable type

> interface Iterable() {

Reviewer 3: ```suggestion
interface Iterable {
```

    let IteratorType:! Iterator;
    fn GetIterator[me: Self]() -> IteratorType;
}

> ```

Reviewer 2: My understanding is that the leads would prefer the iterable type to be involved in the advancing of the iterator.

Author: The `Iterable` type, or the `Iterator`? If we are talking about the `Iterable` (i.e. container), I'm not sure how this is going to work without restricting it to a single global iterator, which does not seem right. Maybe I missed your point?

Author: @josh11b is it something you could clarify? Thanks!

Reviewer 2: This is the cursor approach, I've mentioned on discord. All the methods would be on the `Iterable` type, so there would not need to be any `Iterator` interface. Something like:

```
interface Iterable {
  let ElementType:! Type;
  let CursorType:! Type;
  fn NewCursor[me: Self]() -> CursorType;
  fn GetAtCursor[me: Self](cursor: CursorType) -> ElementType;
  fn AdvanceCursor[me: Self](cursor: CursorType*) -> bool;
}
```
Or:
```
interface Iterable {
  let ElementType:! Type;
  let CursorType:! Type;
  fn NewCursor[me: Self]() -> CursorType;
  fn CursorDone[me: Self](cursor: CursorType) -> bool;
  fn GetAtCursor[me: Self](cursor: CursorType) -> ElementType;
  fn AdvanceCursor[me: Self](cursor: CursorType*);
}
```

An advantage of this approach is that an iterator through an array can do bounds checking in debug mode.

Author: Switched to cursor

These interfaces can be used as such:

```rust
class MyContainer() {
    ...
    external impl as Iterable {
        class IteratorType { // Associated type as member type
            impl as Iterator where .ElementType = i32 {
                ...

                fn Next[addr me: *Self]() -> Optional(ElementType) {
                    ...
                    return if atEnd then .Empty else currItem;
                }
            }
        }
        fn GetIterator[me: Self]() -> IteratorType {
            return IteratorType { ... }
        }
    }
}
```

#### Iterator: `begin()` & `end()`

The solution highlighted embeds the end condition in `Next()`, returns a null
`Optional` at the last elements. This guarantees that `Iterator`s cannot be used
improperly (code safety goal), and exposes a simple interface, but does have a

> performance cost (cost of the `Optional`).

Reviewer: It seems like it could also have an interoperability and migration cost, because Carbon ranged-for loops won't work with C++ containers, and vice-versa.

Reviewer 4: One option for addressing interoperability without binding ourselves to the C++ API would be to have the interop layer automatically inject adopters for language-conforming iterators. e.g., a Carbon `Iterable` could get adapted to look like a C++ iterator for C++ code, whereas a C++ iterator could have an external impl defined for Carbon code (and likely _should_ need that regardless), letting each language keep its own idioms.

(it may make sense to mark interop as an open question since it's not designed out?)

Reviewer: I think it will be challenging to bridge the gap between these two APIs with adapters, because there's a substantial "impedance mismatch" between them that's hard to hide.

- With this design, a Carbon iterator basically plays the same role as a begin/end iterator _pair_ in C++, so there isn't a one-to-one relationship between the objects on either side of the adaptation. That will be a particular problem for automatic adaptation, but may also cause problems for manual adaptation.
- A C++ forward iterator (or better) must support calling `operator*` multiple times, returning the same result so long as the iterator is not advanced. To satisfy that requirement, an adaptor would have to cache the value returned from `Next`.
- C++ iterators support mutation via `operator*`. That compounds the previous problem, because mutation has to apply to the original element, not a cached copy.
- We could avoid those last two problems by saying that Carbon iterators are exposed in C++ as input iterators, but that excludes mutating use cases, and could dramatically degrade performance even for read-only cases, because many algorithms can't be implemented efficiently (or at all) in terms of the input iterator API.

Author: The last points definitely highlight the need for a `Get` function, in place of a simpler `Next`-only approach. I propose to move in that direction.

Regarding the iterator pairs, it is an option to expose an way to get an end iterator for comparison purposes, and interoperability. From a user standpoint, I would probably prefer to not be forced to rely on this end operator, which can be akward at times (needing the container in scope and such), but as a additional feature I believe it makes sense, and likely to have other uses (reverse and such).

**Edit**: Proposal amended with `Get` for `Iterator`.

If we want to retain familiarity with C++ iterators and eliminate this cost, one
option would be to expose `Begin()` and `End()` members functions for
`Iterable`, and allow unsafe access. The lack of pointer arithmetic might hinder
performance improvements for unsafe use, but other solutions may exist.

#### Mutable, Bidirectional, and others

The current `Iterator` interfaces allows simple extension to add features such
as bidirectional access (adding `Prev()` method), or mutable access

> (`GetMutable[addr me: Self*]() -> ElementType*`, or `Set()`)

Reviewer: Should there be a non-mutable `Get` as well? It seems like it would be awkwardly non-orthogonal that if you want to access the element without mutating the iterator, you need to obtain mutable access to the element.

Author: Correct. This should probably be `NextMutable` instead, and doesn't make much sense alone. But based on the interoperability conversation, I will go ahead and add `Get`, which will have the benefit of providing a symmetric solution for this.

    for mutable, other.
-   Maybe
    [Adapting types](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/generics/details.md#adapting-types)
    for flexibility, ad-hoc support for iterating.

    ```
    adapter IterableContainer for Container {
    impl as Iterable { fn Print[me: Self]() { ... } }
    }
    ```

## Rationale

This proposal focuses on user-defined types, which require a great amount of
flexibility with regards to how the final solution is going to be implemented.

In the context, `interface`s represent a non-intrusive, expressive and flexible
solution
([Code that is easy to read, understand, and write](/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)).

It makes use of
[Library APIs only](/docs/project/principles/library_apis_only.md), and is open
to future
[Software and language evolution](/docs/project/goals.md#software-and-language-evolution).

## Alternatives considered

### Using interface type parameters

Interfaces can use
[type parameter](/docs/design/generics/terminology.md#interface-type-parameters-and-associated-types)

> as `interface Iterator(ItemType:! Type` instead of associated types. This would

Reviewer 4: ```suggestion
as `interface Iterator(ItemType:! Type)` instead of associated types. This would
```

range-based `for` loops.

Examples of use cases for `for` loops include iterating over:

-   Mutable and immutable elements
-   Random access containers
-   Maps, hash maps, and complex containers
-   R-value containers
-   Synthetic data
-   Large containers, and large elements
-   Filtering, transforming, and reversing
-   Interoperability with C++ types and containers

The goals for the solution includes:

-   Easy to support for user types, with minimal requirements
-   Simple and clear usage
-   Minimal performance overhead

## Background

The original proposal of `for` loops were discussed in
[p0353](/proposals/p0353.md), and the basic design documented in
[control_flow/loops#for](/docs/design/control_flow/loops.md#for).

The resulting syntax was:

> `for (` _var declaration_ `in` _expression_ `) {` _statements_ `}`

We now need a way to allow supporting this syntax for user and library container
types.

### Other languages

This section highlights some examples of how iterators and range-based for
user-defined types is addressed in different languages.

#### C++

C++ requires us to use the very rich
[`iterator`](https://en.cppreference.com/w/cpp/iterator) library with

> `[c]begin()` and `[c]end()`, to support range-based for.

Reviewer 2: I think the relevant link is: https://en.cppreference.com/w/cpp/language/range-for

The iterator library is not needed. The point is that anything that supports certain methods or function calls may be used.

Author: Thank you, that's the reference I was looking for at the time. Updated.

#### Python

Python requires 2 dunder methods, `__iter__` on the container and `__next__` on

> the iterator, to be usable with a `for element in container:` construct.

Reviewer 2: Maybe also talk about the generator approach for creating an iterator, since that demonstrates how inversion of control works?

Author: Done

    next: u32,
}

impl Iterator for Fibonacci {
    type Item = u32;
    fn next(&mut self) -> Option<Self::Item> {
        // calculate and yield value
        [...]
    }
}

for i in fibonacci().take(4) {
    ...
}
```

#### Typescript

Typescript uses the
[`.iterator` property](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#iterables),
along with
[two forms](https://www.typescriptlang.org/docs/handbook/iterators-and-generators.html#forof-statements)
of range-based for loops: `for..in` and `for..of`.

```ts
let pets = new Set(['Cat', 'Dog', 'Hamster']);
pets['species'] = 'mammals';
for (let pet in pets) {
    console.log(pet); // "species"
}
for (let pet of pets) {
    console.log(pet); // "Cat", "Dog", "Hamster"
}
```

Using the `.iterator` property, Typescript allow to lazily generate data easily
using functions, classes, or objects:

```ts
const reverse = (arr) => ({
    [Symbol.iterator]() {
        let i = arr.length;
        return {
            next: () => ({
                value: arr[--i],
                done: i < 0,
            }),
        };
    },
});
```

#### Go

Go has non-standard support for range-based with custom types only.

> https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-over-a-custom-type

Reviewer 2: I think it would be better to say Go does not officially support customizing `for` for user types (in line with its philosophy of not supporting operator overloading) than refer to this hack.

## Proposal

Provide interfaces that can be implemented by user types to enable support for
ranged-for loops.

We propose:

-   A new `Iterate` interface to support for loops
-   Using a cursor-based approach
-   Making the `for` loop value `let` by default, that is, non-reassignable

Below is a high-level example of this concept:

```carbon
// Add support for `for`
class MyRange {
  impl as Iterate where .ElementType = i32 and .CursorType = i32 {
    // Implement `Iterate`
  }
}
// Usage
let my_range: MyRange = {...};
for (item: auto in my_range) {
  // Do something with `item`
}
```

## Details

### `for` with immutable values

#### `Iterate` interface

This proposal exposes a new `Iterate`. This interface:

-   Relies on a cursor-based approach: minimize implementation effort for
    developers, could support R-values.
-   Expects an `ElementType` and `CursorType`.
-   Combines "advance", "get", and "bounds check" into a single `Next(cursor)`
    method that returns an optional value.

> The proposed `Iterate` is:

Reviewer 2: Try to write this such that it would read naturally once the proposal is accepted, at which point all of this will be true rather than "proposed."

```suggestion
The `Iterate` interface is:
```

```carbon
interface Iterate {
  let ElementType:! type;
  let CursorType:! type;
  fn NewCursor[self: Self]() -> CursorType;
  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType);
}
```

A naive Carbon implementation of the for loop could be:

```carbon
let cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
var iter: Optional(range.(Iterate.ElementType)) = range.(Iterate.Next)(&cursor);

// A. Possible implementation
while (iter.HasValue()) {
  ExecuteForBlock(iter.Get());
  iter = container.(Iterate.Next)(&cursor);
}

// B. Possible alternative depending on the API for Optional(T):
while (true) {
  match (iter) {
    case .None => { break; }
    case .Some(let value: range.(Iterate.ElementType)) => {
      ExecuteForBlock(value);
      iter = container.(Iterate.Next)(&cursor);
    }
  }
}
```

The cursor tracks progression, and the `Next` method advances the cursor and
returns an `Optional` value. An empty `Optional` indicates that we cannot

> advance further.

Reviewer 2: The current wording is a bit unclear to me since it only talks about the "advance" part of `Next` not the "get" part. I'm left not knowing whether there is a final value or not.

```suggestion
returns an `Optional` value. An empty `Optional` indicates that we have reached
the end.
```

Below is a sample implementation of that interface for a user type.

```carbon
class MyIntContainer {
  var data: (i32, i32, i32, i32);
  fn Size[self: Self]() -> i32 {
      return 4;
  }
  impl as Iterate where .ElementType = i32 and .CursorType = i32 {
    fn NewCursor[self: Self]() -> i32 { return 0; }
    fn Next[self: Self](cursor: i32*) -> Optional(i32) {
      // Advance and return value, or return empty
      if (*cursor < self.Size()) {
        let index: i32 = *cursor;
        *cursor = index + 1;

>         return Optional(i32).Create(self.data[index]);

Reviewer 2: Not entirely sure whether this type checks / is legal Carbon. It would depend on there being a special case for tuples that have equal element types.

Author: Agree its definitely not, thanks for highlighting it

      } else {
        return Optional(i32).CreateEmpty();
      }
    }
  }
}
```

Which can be used as follow:

```carbon
fn Main() -> i32 {
  var container: MyIntContainer = {.data = (1, 2, 3, 4)};
  // Implicit `let` value declaration
  for (value: i32 in container) {
    Print("{0}", value);
  }
  return 0;
}
```

#### `let` by default

To keep usage simple and non-surprising, we propose defaulting variable

> declarations to `let` declarations.

Reviewer 2: Say something how this matches what we do in function declarations and match statements (see #2188 ).

```carbon
// Implicitly `let item: T`
for (item: T in my_range) {
  // `item` cannot be reassigned
}
```

This default can be overridden by adding the `var` keyword:

```carbon
for (var item: T in my_range) {

>   // `item` can be modified, but this will not modify `my_range` because it is a copy of the element with its own storage

Reviewer 2: ```suggestion
  // `item` can be modified, but this will not modify `my_range`
  // because it is a copy of the element with its own storage.
```

for (item: T in bucket.GetElements()) {
  // We need to make sure the object is alive for the duration of the `for` loop
}
```

### Use cases

This section covers the use cases highlighted in the introduction, and usage
with the proposed design.

#### Mutable and immutable elements

This proposal covers strictly immutable elements, and proposes defaulting
element declaration to `let`, to minimize surprises and clarify the intent.
Mutable elements are mentioned in the [Future work](#future-work) section.

#### Random access containers

For random access containers, the cursor would represent a numerical index,
incremented within `Next()`.

#### Maps, hash maps, and other containers

Because the cursor type is defined by the developer, and opaque to the user, it
could be a hashmap index, a string, or pointer to a node, without changing the
usage for users.

The `ElementType` can be a tuple, such as a `(key, value)` for maps, or a single
value. See [Future work][#future-work] for other examples.

#### R-value containers

R-values are likely to be limited in terms of addressing and mutation. The
impact is undefined until the

> [corresponding design](https://github.com/carbon-language/carbon-lang/pull/2006/files)

Reviewer 2: ```suggestion
[corresponding design](https://github.com/carbon-language/carbon-lang/pull/2006)
```

This example illustrates a simple proxy interface.

```carbon
class BasicFilter(T:! Iterate) {
  var seq: T*;
  var value: T.ElementType;
  impl as Iterate where .ElementType = T.ElementType and .CursorType = T.CursorType {
    fn NewCursor[self: Self]() -> T.ElementType { return self.seq->NewCursor(); }
    fn Next[self: Self](cursor: T.CursorType*) -> Optional(T.ElementType) {
      var res: auto = self.seq->Next(cursor);
      while (res.has_value() && res.get() == value) {
        res = self.seq->Next(cursor);
      }
      return res;
    }
  }
}

fn Main() -> i32 {
  let container: MyIntContainer = MyIntContainer.Create(0,1,0,2,3,0);
  let no_zeros: BasicFilter(MyIntContainer) =
      { .seq = &container, .value = 0 };
  // Prints "123"
  for (v: i32 in no_zeros) {
    Print("{0}", v);
  }
  return 0;
}
```

On the other hand, changing how the collection is traversed (for example
reverse), requires internal knowledge about the data layout, and a custom
implementation. One approach for containers that support reverse iteration would
be to implement a standard `IterateBidirectional` interface with a `Prev()`

> method. A proxy could be exposed to make it usable with `for` directly.

Reviewer 2: Is the `IterateBidirectional` interface part of the proposal or future work?

Author: Clarified this is an **example**

#### Mutable values

Named mixins, used as programmable members, could expose a view with mutable
values:

```carbon
// Using a named mixin to expose mutable values
for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
```

Used in this way, this clarifies that this a view of the data, like an
attribute, with no direct side effects, when compared to a method.

#### Alternative views

Similarly, mixins could allow supporting other views into the container.

For example, maps could expose a view with "key" and "value" only, in addition
to the default (key, value) tuple.

```carbon
// Using a named mixin to expose mutable values
for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
```

And "reversed" view could also be used for reverse iteration:

```carbon
// Reversing
for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
```

### Large elements and Optional

Large elements would have a negative impact if the `Optional` cannot leverage
unused bit patterns within the type, nor present a view instead of a copy.

Such opportunities for `Optional` exist, and could be addressed in the future.

### C++ interoperability

> The limitation of this approach is the need to adapt to C++ iterator model. This

Reviewer 2: ```suggestion
The limitation of this approach is the need to interoperate with the C++ iterator model. This
```

> could be implemented by having the cursor be an iterator type:

Reviewer 2: There are two directions for interop. I think what you are saying is that we can create a Carbon cursor from a C++ iterator, but I think that should be clarified. It raises the question about interop in the other direction, which should ideally be addressed as well. Maybe there should be two subsections to make it very explicit?

-   `NewCursor` providing the "start iterator"
-   `Next` doing the "increment, bounds check, and returning value"

Two different approaches are identified:

-   One is to have a templated `impl` of `Iterate`, for anything with the right
    method signatures ("begin", "end", etc.)
-   Another is a templated adapter that implements `Iterate` when requested

```carbon
// template impl approach
template constraint CppIterator { ... }
template constraint CppContainer {
  // speculative
  forsome IteratorType: CppIterator;
  fn begin() -> IteratorType;
  fn end() -> IteratorType;
}
external impl forall [template T:! CppContainer] T as Iterate
   where  .CursorType = T.(CppContainer.IteratorType)
      and .ElementType = .CursorType.ElementType {
  fn NewCursor[self: Self]() -> CursorType {
    return self.(CppContainer.begin)();
  }
  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
    if (*cursor == self.(CppContainer.end)()) {
      return Optional(ElementType).MakeEmpty();
    }
    let value: ElementType = **cursor;
    ++*cursor;
    return Optional(ElementType).MakeValue(value);
  }
}
```

```carbon
// adapter approach

> adapter CppIterate(template T:! type) for T {

Reviewer 2: May be worth noting that this approach could have a cursor type parameter (with some sort of default computed from the return type of `T.begin()`), avoiding the deduction issues of the above approach.

Author: Added a mention to that effect.

  impl as Iterate
      // speculative
      where  .CursorType = typeof(T.begin())
        and .ElementType = .CursorType.(Deref.Result) {
    fn NewCursor[self: Self]() -> CursorType {
      return self.begin();
    }
    fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
      if (*cursor == self.end()) {
        return Optional(ElementType).MakeEmpty();
      }
      let value: ElementType = **cursor;
      ++*cursor;
      return Optional(ElementType).MakeValue(value);
    }
  }
}
// used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
```

These two options could allow interoperability with compatible C++ type, either
implicitly (template approach) or explicitly (adapter approach).

### Inversion of control

> Using an Optional has limitations, discussed in the

Reviewer 2: ```suggestion
Using an `Optional` has downsides, discussed in the
```

["alternatives considered"](#alternatives-considered) section. A possible

> approach would be to invert the control of the for loop, by having the container

Reviewer 2: ```suggestion
approach would be to invert the control of the `for` loop, by having the container
```

call in the `for` block for each value, passing the value as argument.

This has the following advantages:

> -   Removes the need for an Optional, associated overheads, copies, and

Reviewer 2: ```suggestion
-   Removes the need for an `Optional`, and the associated overheads, copies, and
```

    unwrapping.
-   No boundary checks needed at the `for` level
-   Compatible with R-value containers

Limitations:

-   The context needed by the `for` body will needs to be available or provided
-   No optimizer known to the team can reliably produce efficient code for this
    construct.

## Rationale

This proposal offers a first step to support a needed way to iterate using a
`for` loop. More specifically:

> -   Requires a single interface to be implemented, uses cursors instead of their

Reviewer 2: ```suggestion
-   Requires a single interface to be implemented with a combined `Next` method,
    uses cursors instead of their
```

    more complex iterator counterpart, and defines loop values as `let` by
    default to be less error prone
    ([Code that is easy to read, understand, and write](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/goals.md#code-that-is-easy-to-read-understand-and-write)).
-   Makes use of
    [Library APIs only](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/project/principles/library_apis_only.md).

## Alternatives considered

### Atomic methods for `Iterate`

An option was to define methods such as `Get`, `Advance`, `IsAtEnd` on
`Iterate`, instead of a unique `Next` methods. While being more atomic, the
concern is that these individual methods will all need to be called separately
while iterating, while a single call to `Next` performs these three operations
together, providing more opportunities for optimization.

Using a single operation to return an optional is also consistent with Rust’s
approach.

> The drawback of this `Next` method is the fact it returns an `Optional`, though

Reviewer 2: This paragraph would be clearer if it stated what the downsides of returning an optional was.

Author: Changed the way this is presented, does this look better to you?

ways to limit or eliminate that cost seem to exist (value reuse or view for
Optional, inversion of control on `for` block statement).

> ### Returning `T` instead of `Optional(T)`
> 
> Using `Optional(T)` is a side effect of combining operations, discussed in the
> previous section.
> 

Reviewer 2: I don't think this section is adding any additional value beyond the previous section.

```suggestion
```

### Using an iterator instead of a cursor

The iterator approach was considered but proved to have key limitations that a
cursor approach does not have:

-   It requires implementing 2 interfaces instead of 1, and is more complex due
    to having the iteration logic separated from the container itself
-   More difficult to harden or troubleshoot, compared to a cursor that allows
    bounds checking in debug & hardened build modes
-   Higher overhead
-   Proves to be difficult to support for R-values

On the other hand, the cursor approach deviates more from C++ than iterator, and
may require some more effort for C++ interoperability.

> ### Single interface for immutable and mutable data

Reviewer 2: Clarify that this is talking about not having methods for handling both immutable and mutable data in one interface. Right now we are talking about having a single interface, and using it with `T` for immutable data, and `T*` for mutable access.

Author: Done

Because some collections will be read-only, we want to separate mutable and

> immutable interfaces to opt-in only on demand. A single interface would force developers

Reviewer 2: ```suggestion
immutable interfaces. A single interface would force developers
```

```carbon
// Using a named mixin to expose mutable values
for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
```

And "reversed" view could also be used for reverse iteration:

```carbon
// Reversing
for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
```

### Large elements and Optional

Large elements would have a negative impact if the `Optional` cannot leverage
unused bit patterns within the type, nor present a view instead of a copy.

Such opportunities for `Optional` exist, and could be addressed in the future.

### C++ interoperability

The limitation of this approach is the need to adapt to C++ iterator model. This
could be implemented by having the cursor be an iterator type:

-   `NewCursor` providing the "start iterator"
-   `Next` doing the "increment, bounds check, and returning value"

Two different approaches are identified:

-   One is to have a templated `impl` of `Iterate`, for anything with the right
    method signatures ("begin", "end", etc.)
-   Another is a templated adapter that implements `Iterate` when requested

```carbon
// template impl approach
template constraint CppIterator { ... }
template constraint CppContainer {
  // speculative

>   forsome IteratorType: CppIterator;

Reviewer 2: Having thought about it more, we need this type to be unique if it is going to be deduced, which is a requirement for using its value (which we do when writing `T.(CppContainer.IteratorType)` below). So it would be better if this speculative syntax used `for_unique` instead of `forsome`:
```suggestion
  for_unique IteratorType: CppIterator;
```

  fn begin() -> IteratorType;
  fn end() -> IteratorType;
}
external impl forall [template T:! CppContainer] T as Iterate

>    where  .CursorType = T.(CppContainer.IteratorType)

Reviewer 2: ```suggestion
    where .CursorType = T.(CppContainer.IteratorType)
```

      and .ElementType = .CursorType.ElementType {
  fn NewCursor[self: Self]() -> CursorType {
    return self.(CppContainer.begin)();
  }
  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
    if (*cursor == self.(CppContainer.end)()) {
      return Optional(ElementType).MakeEmpty();
    }
    let value: ElementType = **cursor;
    ++*cursor;
    return Optional(ElementType).MakeValue(value);
  }

> }

Reviewer 2: ```suggestion
}
// used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v) { ... }
```

```

```carbon
// adapter approach
adapter CppIterate(template T:! type) for T {
  impl as Iterate
      // speculative

>       where  .CursorType = typeof(T.begin())

Reviewer 2: ```suggestion
      where .CursorType = typeof(T.begin())
```

        and .ElementType = .CursorType.(Deref.Result) {
    fn NewCursor[self: Self]() -> CursorType {
      return self.begin();
    }
    fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
      if (*cursor == self.end()) {
        return Optional(ElementType).MakeEmpty();
      }
      let value: ElementType = **cursor;
      ++*cursor;
      return Optional(ElementType).MakeValue(value);
    }
  }
}
// used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
```

> These two options could allow interoperability with compatible C++ type, either

Reviewer 2: ```suggestion
These two options could allow interoperability with a compatible C++ type, either
```

We propose:

-   A new `Iterate` interface to support for loops
-   Using a cursor-based approach
-   Making the `for` loop value `let` by default, that is, non-reassignable

Below is a high-level example of this concept:

```carbon
// Add support for `for`
class MyRange {
  impl as Iterate where .ElementType = i32 and .CursorType = i32 {
    // Implement `Iterate`
  }
}
// Usage
let my_range: MyRange = {...};
for (item: auto in my_range) {
  // Do something with `item`
}
```

## Details

### `for` with immutable values

#### `Iterate` interface

This proposal exposes a new `Iterate`. This interface:

-   Relies on a cursor-based approach: minimize implementation effort for
    developers, could support R-values.
-   Expects an `ElementType` and `CursorType`.
-   Combines "advance", "get", and "bounds check" into a single `Next(cursor)`
    method that returns an optional value.

The `Iterate` interface is:

```carbon
interface Iterate {
  let ElementType:! type;
  let CursorType:! type;
  fn NewCursor[self: Self]() -> CursorType;
  fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType);
}
```

A naive Carbon implementation of the for loop could be:

```carbon

> let cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();

Reviewer 5: if the curser is mutated during the iteration, shouldn't it be `var curser : ...`?

Reviewer 2: You are correct. Good catch!

```suggestion
var cursor: range.(Iterate.CursorType) = range.(Iterate.NewCursor)();
```

Author: Thanks!

-   Facilitate code reuse, compared to reimplementing an interface
-   Direct access to `self`, limiting needs for pointers and address resolution

#### Mutable values

Named mixins, used as programmable members, could expose a view with mutable
values:

```carbon
// Using a named mixin to expose mutable values
for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
```

Used in this way, this clarifies that this a view of the data, like an
attribute, with no direct side effects, when compared to a method.

#### Alternative views

Similarly, mixins could allow supporting other views into the container.

For example, maps could expose a view with "key" and "value" only, in addition
to the default (key, value) tuple.

```carbon
// Using a named mixin to expose mutable values
for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
```

And "reversed" view could also be used for reverse iteration:

```carbon
// Reversing
for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
```

### Large elements and Optional

Large elements would have a negative impact if the `Optional` cannot leverage
unused bit patterns within the type, nor present a view instead of a copy.

> Such opportunities for `Optional` exist, and could be addressed in the future.

Reviewer: I'm skeptical that there are opportunities for `Optional(T)` to present a view of the `T` object instead of a copy. Currently, Carbon has no way of expressing that, and there is substantial resistance to adding a way to do so, because a view of an arbitrary type is basically a reference. As far as I can tell, if we want to solve this problem without introducing references, we have to do it as part of the design of `for`, not the design of `Optional`. That way we can define the reference-like behavior as part of the hidden machinery that `for` rewrites to, without giving user code access to the full power of reference types. See #2274 for an example of something similar, where we enabled a particular kind of reference-like behavior by defining special rules for how `[]` is rewritten to a method call.

In many ways inversion of control would be an ideal solution here; it's really unfortunate that it's apparently so hard to optimize.

Author: If I understand that correctly, you'd suggest having a similar rewrite that bypasses `Optional`s entirely. Which means needing to expose an `Optional`-less API (but then losing the advantage of having a single `Next` method), or short-circuiting the `return` before the Optional is created (is that possible?). I'm not sure the same situation applies exactly, feel free to correct me.

I was skeptical too, @zygoloid seemed somewhat confident about it in a previous discussion. But I haven't done a good job at capturing here what he had in mind. @zygoloid would you mind sharing again what you had in mind if you recall?

Reviewer: I'm not suggesting bypassing `Optional` entirely; I'm only saying that I don't expect `Optional` to solve the problem of providing view-like performance, and so it will need to be solved as part of the design of `for`.

We could solve this by having `Next()` return `Optional(T*)`, except that doesn't work on its own because the elements of the underlying range might not be lvalues (either because the range itself is an rvalue or because the range type does something like compute elements on the fly). So sometimes we want a next-element operation that returns `Optional(T)`, and other times we want one that returns `Optional(T*)`. That all seems very similar to the problem that we had to solve with subscripting, where sometimes we want a get-nth-element operation that returns `T`, and other times we want one that returns `T*`, so it seems like a similar solution might work here. I think this would also let us support mutating the elements being iterated over, which is a nice bonus.

Reviewer: We discussed this a bit at the weekly meeting today, and a couple of other options came up:
- We could let the container choose whether it exposes values or pointers to values, and maybe provide a dereferencing adapter that turns a range of pointers into a range of values. The nice thing is this doesn't involve any technical changes to the design, just a shift in expectations about how containers will typically implement `Iterate`. However, that seems pretty awkward from a user experience point of view, and I think it would mean that any given container type could support efficient iteration, or iteration over container r-values, but not both, so I think this only makes sense as a stopgap while we investigate a longer-term solution like control inversion.
- We could allow the binding in the `for` loop be an `addr` pattern, which would be implemented by calling into a separate interface on the container. This puts the choice completely in the hands of the `for` loop author, so it wouldn't let us match the ergonomics of `let` (where the compiler can choose to copy or alias the value depending on what's more efficient), but it at least gives us parity with C++ (where the author has to choose whether the iteration variable is a reference).

Author: Thank you, updated to reflect the options suggested here, and noted that the preference today seems to be with #1, to get to a first step before revising.

      return Optional(ElementType).MakeEmpty();
    }
    let value: ElementType = **cursor;
    ++*cursor;
    return Optional(ElementType).MakeValue(value);
  }
}
// Used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v) { ... }
```

```carbon
// Adapter approach
adapter CppIterate(template T:! type) for T {
  impl as Iterate
      // Speculative syntax. Alternatively, a cursor type parameter
      // would avoid the need for deduction.
      where .CursorType = typeof(T.begin())
        and .ElementType = .CursorType.(Deref.Result) {
    fn NewCursor[self: Self]() -> CursorType {
      return self.begin();
    }
    fn Next[self: Self](cursor: CursorType*) -> Optional(ElementType) {
      if (*cursor == self.end()) {
        return Optional(ElementType).MakeEmpty();
      }
      let value: ElementType = **cursor;
      ++*cursor;
      return Optional(ElementType).MakeValue(value);
    }
  }
}
// Used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
```

These two options could allow interoperability with a compatible C++ type,
either implicitly (template approach) or explicitly (adapter approach).

> #### Supporting Carbon `Iterate` in C++

Reviewer: Does this mean "Implementing the `Iterate` interface in C++ code", or "Implementing C++ iteration in terms of `Iterate`"? The phrasing sounds more like the first, but from context I think you mean the second.

Maybe the previous section should be called "Iterating over C++ types in Carbon", and this section should be called "Iterating over Carbon types in C++".

Author: Agree that your suggestions are clearer, thanks!

Support for `Iterate` in C++ needs to provide `begin()` and `end()` methods or
free functions, in addition to `*` and `++` operators.

`Iterate` does not expose a way to have a cursor or iterator to the past-last
element. An option would be to have `end()` return a predefined invalid value,
to satisfy iterator comparison when `Next()` returns an empty `Optional`.

Below is an incomplete example of classes exposing those methods:

> ```cpp

Reviewer: I'm having difficulty following this example -- is `Cursor` a cursor, or an iterator that wraps a cursor? How does `Cursor` relate to `CppIterate`? What do the template parameters of `CppIterate` mean?

Apart from those API questions, I think it would be very helpful to flesh out the implementation of `Cursor` in more detail, showing the data members and at least some of the method bodies. That would help answer questions like:

- How do we support calling `operator*` twice on the same `Cursor`?
- Assuming that `Cursor` is meant to be a C++ iterator, how do we support copying it?

Author: Added a complete (working) implementation. Might not be the best way to do it, but this gives a first idea of what is necessary.

template <typename T>
class Cursor {
  Cursor<T> FromCarbonCursor(...);
  Cursor<T> Invalid();

  T& operator*() const;
  Cursor<T>& operator++();
  // `true` if both are equal or invalid/empty
  bool operator==(const Cursor<T>& rhs);
  bool operator!=(const Cursor<T>& rhs);
  // [...]
};

template <typename C, typename T>
class CppIterate {

>   C start() { return C::FromCarbonCursor(iterate_.NewCursor()); }

Reviewer: ```suggestion
  C begin() { return C::FromCarbonCursor(iterate_.NewCursor()); }
```

pattern, implemented by calling into a separate interface on the container. This
puts the choice of using values or pointers to values in the hands of the `for`
loop author. While this deviates from the ergonomics of `let` (where the
compiler can choose to copy or alias the value depending on what is more
efficient), this option would provide parity with C++ `for` loops (where the
author chooses between value and reference).

### C++ interoperability

#### Iterating over C++ types in Carbon

For reference, range-based `for` loops in C++ requires:

-   `begin()` and `end()` methods or free functions
-   supporting pre-increment operator `++` and indirection operator `*`

See section 8.6.4 of N4860 for more details.

A limitation of the approach proposed in this document is the need for
adaptation to interoperate with the C++ iterator model. This could be
implemented by having the cursor be an iterator type:

-   `NewCursor` providing the "start iterator"
-   `Next` doing the "increment, bounds check, and returning value"

Two different approaches are identified:

-   One is to have a templated `impl` of `Iterate`, for anything with the right

>     method signatures ("begin", "end", etc.)

Reviewer 2: ```suggestion
    method signatures (`begin`, `end`, and so on)
```

    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
  }
  auto end() -> IterateIterator<Iterable<C, T>> {
    return IterateIterator<Iterable<C, T>>::Invalid(*this);
  }
};
}  // namespace Carbon

// Usage:
Carbon::MyContainer container;
for (auto s : container) { ...}
```

### Inversion of control

Using an `Optional` has downsides, discussed in the
["alternatives considered"](#alternatives-considered) section. A possible
approach would be to invert the control of the `for` loop, by having the
container call in the `for` block for each value, passing the value as argument.
This would work similarly to [Python generator functions](#python).

This has the following advantages:

-   Removes the need for an `Optional`, and the associated overheads, copies,
    and unwrapping.
-   No boundary checks needed at the `for` level
-   Compatible with R-value containers

Limitations:

-   The context needed by the `for` body will needs to be available or provided
-   No optimizer known to the team can reliably produce efficient code for this
    construct.

## Rationale

This proposal offers a first step to support a needed way to iterate using a
`for` loop. More specifically:

> -   Requires a single interface to be implemented with a combined `Next` method,

Reviewer: It might also be worth mentioning that a combined `Next` method makes it easier to support ranges that consume their input. For example, in C++20 if you filter an `istream_view` through `take(5)` (which filters a range to its first 5 elements), the first *6* elements of the stream are consumed, because the C++ iterator model forces `istream_iterator` to consume an element on initialization (so that the first `operator*` has a result to return) and on every `operator++` call. As another example of the same kind of problem, `iota(0) | filter([](auto x) { return x < 5; })) | take(5)` is an infinite loop. I think a combined `Next` method will avoid these problems.

I got this from a talk by Barry Rezvin at C++Now last week; I'm not sure when the video will be online, but you can see the slides [here](https://schedule.cppnow.org/wp-content/uploads/2023/03/take5-slides.pdf)

Author: Added to the "Alternatives" section, as this seemed to best belong there.

uses speculating on the final design. Some may include:

-   Improved semantics for views compared to getter methods: no direct side
    effect from using the view
-   Facilitate code reuse, compared to reimplementing an interface
-   Direct access to `self`, limiting needs for pointers and address resolution

#### Mutable values

Named mixins, used as programmable members, could expose a view with mutable
values:

```carbon
// Using a named mixin to expose mutable values
for (p: my_range.(Iterate.ElementType)* in my_range.mutable) {}
```

Used in this way, this clarifies that this a view of the data, like an
attribute, with no direct side effects, when compared to a method.

#### Alternative views

Similarly, mixins could allow supporting other views into the container.

For example, maps could expose a view with "key" and "value" only, in addition
to the default (key, value) tuple.

```carbon
// Using a named mixin to expose mutable values
for (k: my_dict.(Iterate.CursorType) in my_dict.keys) {}
```

And "reversed" view could also be used for reverse iteration:

```carbon
// Reversing
for (v: my_dict.(Iterate.ElementType) in my_dict.reversed) {}
```

> ### Large elements and Optional

Reviewer 2: @chandlerc I feel like we had a conversation about this and r-value representations. Could you recall where we ended up with this?

```carbon
adapter IterateByValue(T:! Iterate where .CursorType impls Deref) for T {
  impl as Iterate
      where .CursorType = T.CursorType
        and .ElementType = T.CursorType.(Deref.Result) {
        ...
  }
}

// Used like:
var frames: NetworkFramesContainerByPtr = ...
for (i: NetworkFrame in frames as IterateByValue(NetworkFramesContainerByPtr))
{
  ...
}
```

Alternatively, we can allow the binding in the `for` loop to be an `addr`
pattern, implemented by calling into a separate interface on the container. This
puts the choice of using values or pointers to values in the hands of the `for`
loop author. While this deviates from the ergonomics of `let` (where the
compiler can choose to copy or alias the value depending on what is more
efficient), this option would provide parity with C++ `for` loops (where the
author chooses between value and reference).

### C++ interoperability

#### Iterating over C++ types in Carbon

For reference, range-based `for` loops in C++ requires:

-   `begin()` and `end()` methods or free functions, and
-   the type returned supports pre-increment `++`, indirection `*`, and
    inequality `!=` operations

> See section 8.6.4 of N4860 for more details.

Reviewer 2: Can this be made into a link?

Author: Do we have any preferred source to link the standard in an easy to read way? This is what I have https://eel.is/c++draft/stmt.iter#stmt.ranged from the latest draft.

Reviewer 2: I don't have a preferred source. The link you mention is fine with me. For non-spec C++ links, I tend to use cppreference.com

Author: Done

    -   [Other interfaces](#other-interfaces)
        -   [Compatibility](#compatibility)
        -   [Possible interface for 'Consumable' containers](#possible-interface-for-consumable-containers)
            -   [Range-based `for` behavior](#range-based-for-behavior-1)
            -   [Interface and usage](#interface-and-usage-1)
    -   [Limitations and future language improvements](#limitations-and-future-language-improvements)
-   [Rationale](#rationale)
-   [Alternatives considered](#alternatives-considered)
    -   [Indexable](#indexable)
    -   [Using interface type parameters](#using-interface-type-parameters)
    -   [Relying only on `Next()` from `Iterator` to move and get value](#relying-only-on-next-from-iterator-to-move-and-get-value)
    -   [Iterator's `begin()` & `end()` methods](#iterators-begin--end-methods)
    -   [Returning `Optional(T)` instead of `T` for getters](#returning-optionalt-instead-of-t-for-getters)
    -   [Other](#other)

<!-- tocstop -->

## Abstract

The goal of this proposal is to provide a way for user-defined types to support
range-based iteration with `for`. The proposed solution exposes an interface

> that can be implemented by user types to enable support for ranged-`for` loops.

Reviewer 2: Feels like this could be a little more concise.

```suggestion
This proposes an interface that can be implemented by user types to support
range-based iteration with `for`.
```

## Problem

The current `for` proposal does not define a way for user-defined types to
support range-based for loops. Adding support would make it significantly easier
to define custom containers, as well as containers in Carbon standard library.

Example of data and use cases include:

-   Iterating over mutable and immutable data

> -   Iterating over r-value container

Reviewer 2: ```suggestion
-   Iterating over an r-value container
```

opt-in to support range-based for loops. An approach is proposed in this
document.

> This can be achieved using either a cursor or an iterator. A cursor benefits

Reviewer: It's probably worth spelling out the difference between "cursor" and "iterator"; I'm not sure everyone is familiar with that terminology.

#### C++

C++ requires either `.begin()`/`.end()` methods, or
`begin(<container>)`/`end(<container>)` to be supported for
[range-based `for`](https://en.cppreference.com/w/cpp/language/range-for).

#### Python

Python requires 2 dunder methods, `__iter__` on the container and `__next__` on
the iterator, to be usable with a `for <element> in <container>:` construct.

A lazy iterator can also be returned using `yield`, giving directly control over

> the values presented for each iteration.

Reviewer 2: ```suggestion
[Generator functions](https://peps.python.org/pep-0255/) act like iterators.
They are executed until the next `yield` statement, whose argument is returned
as the next value to the `for` loop. Iteration ends when the function returns.
```

```python

> def generate():

Reviewer 2: ```suggestion
def generate_fibonacci():
```

Provide interfaces that can be implemented by user types to enable support for
ranged-for loops.

We propose a new [**Iterable**](#iterable) interface, for iterator-based
sequential access, and usage with range-based `for`.

## Details

This proposal exposes an Iterable-type interface, relying on 2 new interfaces:
`Iterable` and `Iterator`.

A user implements these interfaces to enable range-based `for` support. Carbon
would make use of the `Iterable` to retrieve an `Iterator` and iterate until
completed or interrupted.

### Iterable

The `Iterable` interface represents a generic sequential-access type of
container. It relies on two new interfaces:

-   `Iterable` and `MutableIterable`: Interface for a type that supports being
    iterated over
-   `Iterator`: Interface for an iterator-like object

#### Range-based `for` behavior

-   Get an `Iterator` type with `GetIterator()` or `GetMutableIterator()`
-   Call `Advance() -> bool`
    -   Break the loop if the result is `false`
-   Call `Get()` and return the result

#### Interface and usage

These interfaces are defined as:

```carbon
// Basic immutable iterator

> interface Iterator {

Reviewer 2: Why go with the iterator approach instead of the cursor approach? I think the cursor approach has a number of advantages, and is what I've heard @chandlerc prefers. For example, I think the cursor approach supports iterating through r-value arrays.

Reviewer: Iterators are likely to interop better with C++. But yes, cursors should definitely at least be an "alternative considered".

IMO we should really avoid statements like "I've heard @chandlerc prefers it".
- It's an argument from authority.
- Citing one lead in that way tends to undermine the other two.
- It's (at least) third-hand information, so there's a high risk of information getting lost, and there's no way to engage with it constructively.

Reviewer 2: I'm sorry, it was not my intention to make an argument from authority. I was trying to say something about how I had heard different stories in different venues from different people and I was hoping they would connect to discuss it. My understanding of the goals of the proposal process is to gather and represent the different opinions and approaches, ideally before the proposal is sent to the leads for final ratification and I perceived a gap. Unfortunately, my attempt to add @chandlerc to the conversation did not lead to that outcome. Next time I will try harder to make it clear that this was an invitation to continue a conversion rather than attempt to end it.

Author: Moved to cursor approach following discussion with leads.

Large elements would have a negative impact if the `Optional` cannot leverage
unused bit patterns within the type, nor present a view instead of a copy.

If this proves difficult, we can let implementers of the container choose
whether to expose values or pointers to values. When electing to "pointers to
values", an adapter can be provided to support transforming the range of
pointers into a range a values. This choice allows supporting either iteration
over container r-values (`ElementType = T`), or efficient iteration
(`ElementType = T*`), but not both. While acceptable as an intermediate
solution, a better alternative will be needed in the future.

```carbon

> adapter IterateByValue(template T:! type) for T {

Reviewer 2: I think this can be made into a checked generic instead of a template if we can find the right constraint:

```suggestion
adapter IterateByValue(T:! Iterate where .CursorType impls Deref) for T {
```

Author: Done, I cannot confirm if this syntax is correct, but can rely on others to mention it if needed.

  impl as Iterate where
      // Speculative syntax. Alternatively, a cursor type parameter
      // would avoid the need for deduction.
      where .CursorType = T.CursorType

>         and .ElementType = T.CursorType.(Deref.Result) {

Reviewer 2: With the checked generic parameter, I don't think there is anything speculative here, nor anything calling for a cursor type parameter.
```suggestion
  impl as Iterate
      where .CursorType = T.CursorType
        and .ElementType = T.CursorType.(Deref.Result) {
```

Author: My bad, though the `Deref` syntax was, but seems like it exists in the doc. Done.

        ...
  }
}

// Used like:
var frames: NetworkFramesContainerByPtr = ...
for (i: NetworkFrame in frames as IterateByValue(NetworkFramesContainerByPtr))
{
  ...
}
```

Alternatively, we can allow the binding in the `for` loop to be an `addr`
pattern, implemented by calling into a separate interface on the container. This
puts the choice of using values or pointers to values in the hands of the `for`
loop author. While this deviates from the ergonomics of `let` (where the
compiler can choose to copy or alias the value depending on what is more
efficient), this option would provide parity with C++ `for` loops (where the
author chooses between value and reference).

### C++ interoperability

#### Iterating over C++ types in Carbon

For reference, range-based `for` loops in C++ requires:

-   `begin()` and `end()` methods or free functions

> -   supporting pre-increment operator `++` and indirection operator `*`

Reviewer 2: Two things:
- Something to connect these two requirements to make it clear that the type returned by `begin()` and `end()` is what has to support these operators.
- I believe the iterator type must also support inequality comparison (at least). According to https://en.cppreference.com/w/cpp/language/range-for a C++ range-based `for` loop gets rewritten to include `for ( ; __begin != __end; ++__begin)`.

We may need to be careful about closely matching C++ so we can be as compatible as we can manage, as long as it doesn't impact the usability for ordinary users. In C++, my understanding is as long as the rewrite type checks, the ranged-based for loop is legal, like a template. This means we may need to support `begin()` and `end()` returning different types, as long as `!=` is defined on those two types.

Author: Absolutely yes, clarified.

See section 8.6.4 of N4860 for more details.

> A limitation of the approach proposed in this document is the need for
> adaptation to interoperate with the C++ iterator model. This could be
> implemented by having the cursor be an iterator type:

Reviewer 2: ```suggestion
A limitation of the approach proposed in this document is the need for
adaptation to interoperate with the C++ iterator model. The adapter would
be a Carbon type that satisfies the Cursor interface, implemented in terms
of a C++ iterator.
```

> -   `NewCursor` providing the "start iterator"

Reviewer 2: ```suggestion
-   `NewCursor` providing the "start iterator" returned by `begin()`
```

    }
  }
}
// Used like:
var v: Cpp.std.vector(i32) = ...
for (i: i32 in v as CppIterate(Cpp.std.vector(i32))) { ... }
```

These two options could allow interoperability with a compatible C++ type,
either implicitly (template approach) or explicitly (adapter approach).

### Inversion of control

Using an `Optional` has downsides, discussed in the
["alternatives considered"](#alternatives-considered) section. A possible
approach would be to invert the control of the `for` loop, by having the

> container call in the `for` block for each value, passing the value as argument.

Reviewer 2: ```suggestion
container call in the `for` block for each value, passing the value as argument.
This would work similarly to [Python generator functions](#python).
```

These two options could allow interoperability with a compatible C++ type,
either implicitly (template approach) or explicitly (adapter approach).

#### Iterating over Carbon types in C++

> C++ needs to provide `begin()` and `end()` methods or free functions, in
> addition to `*` and `++` operators.

Reviewer 2: ```suggestion
There need to be `begin()` and `end()` methods or free functions accessible to
C++ for any Carbon type that implements the `Iterate` interface, in
addition to overloads of the `*`, `++`, and `!=` operators for the types returned.
```

Author: Great, thanks for clarifying.

invalid value, to satisfy iterator comparison when `Next()` returns an empty
`Optional`.

Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
need to copy and cache the last value, to allow for `*it` to be called. This
puts a requirement on `ElementType` to be copyable for this interoperability to
be usable.

```cpp
namespace Carbon {
// A C++ input iterator for any type that implements the Carbon `Iterate` interface.
template <typename Iterate>
class IterateIterator {
 public:
  // Returns an "Invalid" iterator representing `end()`
  static auto Invalid(const Iterate& container) -> IterateIterator<Iterate> {
    return IterateIterator<Iterate>(container, std::nullopt);
  }

  IterateIterator(const Iterate& container,
                  std::optional<typename Iterate::CursorType> cursor)
      : container_(&container), cursor_(cursor) {
    if (cursor_) {
      next();
    }
  }
  IterateIterator(const IterateIterator& other)
      : container_(other.container_), cursor_(other.cursor_),
      value_(other) {}

  auto operator*() const -> const typename Iterate::ElementType& {
    assert(cursor_);
    return value_;
  }

>   auto operator++() -> IterateIterator<Iterate>& {

Reviewer 2: Probably should have both this and the version that takes an `int` parameter, so this iterator supports both pre- and post- increment.

Reviewer: I don't think I agree. Post-increment won't work if `ElementType` isn't copyable. By the same token, this type won't be a valid iterator if `ElementType` isn't copyable, but it will still work with range-based for loops, so I think we should focus solely on the features needed to support that.

Author: Agree with @geoffromer , resolving this one as I don't think this would be a blocking issue. Feel free to re-open if you think otherwise @josh11b .

    next();
    return *this;
  }
  auto operator==(const IterateIterator<Iterate>& rhs) -> bool {
    return container_ == rhs.container_ && cursor_ == rhs.cursor_;
  }
  auto operator!=(const IterateIterator<Iterate>& rhs) -> bool {
    return !(*this == rhs);
  }

 private:
  void next() {
    assert(cursor_);
    if (const auto v = container_->Next(*cursor_); v) {
      value_ = *v;
    } else {
      cursor_ = std::nullopt;
    }
  }

  const Iterate* container_;
  std::optional<typename Iterate::CursorType> cursor_;
  typename Iterate::ElementType value_;
};

// A type implementing `Iterate` with CursorType `C`, and
// ElementType `T` could have `begin()` and `end()` methods
// defined as such:
class IterateType ... {
  auto begin() -> IterateIterator<Iterable<C, T>> {
    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
  }
  auto end() -> IterateIterator<Iterable<C, T>> {

>     return IterateIterator<Iterable<C, T>>::Invalid(*this);

Reviewer 2: What is `Iterable<...>` here? It looks like `IterateIterator` expects to be parameterized by a container type. Is `Iterable` some sort of adapter to define the `CursorType` and `ElementType` members? I don't see it defined anywhere else in this file.

Author: That's a typo, and should be  `Iterate<C, T>`, which is the C++ equivalent to Carbon's type. Fixed.

These two options could allow interoperability with a compatible C++ type,
either implicitly (template approach) or explicitly (adapter approach).

#### Iterating over Carbon types in C++

C++ needs to provide `begin()` and `end()` methods or free functions, in
addition to `*` and `++` operators.

The `Iterate` interface does not expose a way to have a cursor nor an iterator
to the past-last element. An option would be to have `end()` return a predefined
invalid value, to satisfy iterator comparison when `Next()` returns an empty
`Optional`.

Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
need to cache the last value, to allow for `*it` to be called, and to be called

> more than once.

Reviewer: It seems worth calling out that this will mean that the iterator is not copyable if the value type is not copyable. That's not a problem as far as C++'s ranged-for is concerned, but it's a problem for the C++ standard library, which requires all iterators to be copyable.

Author: This should be clearer now

```cpp
#include <cassert>
#include <iostream>
#include <optional>
#include <string>

namespace Carbon {

template <typename C, typename T>

> class Iterable;

Reviewer: Does this forward declaration get used anywhere in `IterateIterator`?

Author: Not needed anymore

> template <typename Iterable>

Reviewer: I'm guessing that the template argument can be the C++ counterpart of any Carbon type that implements `Iterate`, right? If so, calling it `Iterable` instead of `Iterate` is confusing. Same goes for the `Iterable` template, and the `MyIterable` class.

Also, I think a comment would make this a lot clearer, e.g.

```suggestion
// A C++ input iterator for any type that implements the Carbon `Iterate` interface.
template <typename Iterable>
```

Author: Right, I battled that a minute, intending to use `Iterable` as a kind of concept, rather than the actual `Iterate`. Reads better, but agree its a bit more confusing.

  }
  auto operator!=(const IterateIterator<Iterable>& rhs) -> bool {
    return !(*this == rhs);
  }

 private:
  void next() {
    assert(cursor_);
    if (const auto val = container_->Next(*cursor_); val) {
      value_ = *val;
    } else {
      cursor_ = std::nullopt;
    }
  }

  const Iterable* container_;
  std::optional<typename Iterable::CursorType> cursor_;
  typename Iterable::ElementType value_;
};

template <typename C, typename T>
class Iterable {
 public:
  using CursorType = C;
  using ElementType = T;

  virtual auto NewCursor() const -> CursorType = 0;
  virtual auto Next(CursorType& cursor) const -> std::optional<ElementType> = 0;

  auto begin() -> IterateIterator<Iterable<C, T>> {
    return IterateIterator<Iterable<C, T>>(*this, NewCursor());
  }
  auto end() -> IterateIterator<Iterable<C, T>> {
    return IterateIterator<Iterable<C, T>>::Invalid(*this);
  }
};

class MyIterable : public Iterable<int, std::string> {
  auto NewCursor() const -> CursorType override { return 0; }
  auto Next(CursorType& cursor) const -> std::optional<ElementType> override {
    if (cursor < 10) {
      return std::to_string(cursor++);
    }
    return std::nullopt;
  }

> };

Reviewer: I think it might be clearer to omit `Iterable` and `MyIterable`, and instead have text saying that the generated C++ counterpart for any Carbon type that implements `Iterate` will have `begin` and `end` methods defined as follows, and then show those methods in isolation. The rest of this code doesn't really address the key issues for this proposal, and raises other questions that aren't important here (e.g. will we really use inheritance?)

Author: Great suggestion, thanks.

> };  // namespace Carbon

Reviewer: ```suggestion
}  // namespace Carbon
```

Using a single operation to return an optional is also consistent with Rust’s
approach.

> The combined `Next` method forces us to return an `Optional`, which has a

Reviewer 2: ```suggestion
The combined `Next` method forces us to return an `Optional`, which has
```

downsides:

> -   Potential performance overhead to wrapping and unwrapping, and storage
>     overhead for `Optional` itself
> -   May require copying the value wrapped

Reviewer 2: ```suggestion
-   potential performance overhead to wrapping and unwrapping,
-   storage overhead for `Optional` itself, and
-   may require additional copying of the value wrapped.
```

Those may be mitigated by leveraging unused bit patterns, passing a view of the

> value, or inverting control for of `for` block statements.

Reviewer 2: ```suggestion
value, or [inverting control](#inversion-of-control).
```

`Optional`.

Below is sample implementation to iterate on a Carbon `Iterate`. Note that we
need to copy and cache the last value, to allow for `*it` to be called. This
puts a requirement on `ElementType` to be copyable for this interoperability to

> be usable.

Reviewer: I think this will still work if `ElementType` is move-only (but see the comment below), it just means that `IterateIterator` will be move-only in those cases. That means it won't be a valid iterator according to the standard library, but it should still work with range-based `for` loops.

Author: I just looked back at the code: 
```
     if (const auto v = container_->Next(*cursor_); v) {
      value_ = *std::move(v);
    }
```
Won't `Next()` require `ElementType` to be copyable nonetheless, as we are effectively returning a copy of an element, and not moving anything from the container? Even if we avoid an optional-like wrapper, or just move from it, the `ElementType` still has to be copied at some point (likely withing `Next()`).

Maybe I forgot something between now and then.

Reviewer: In recent versions of C++, if `Next` returns an `ElementType` rather than a reference, the code you quote won't actually copy or move the return value; it will use `v` itself as the storage for the returned temporary. Now, you're right that in this context, the _body_ of `Next` will very likely need to make a copy, but `Next` is written in pure Carbon, so that problem isn't caused by caching in the interoperability layer, and in fact it's not caused by the interop layer at all -- the same problem exists in pure Carbon.

Author: Agree that this is by design, and not caused by the interop. To expand, I think we could say "move-only" in the C++ domain, most that would be ignoring a very likely "copiable" requirement on Carbon side / `Next`. So all-in-all, "copyable" seem more correct, and the upper bound of reqs, vs the lower bound.

Reviewer: The problem is that this doc doesn't explicitly talk about copyability on the Carbon side, so by introducing it here as a new topic that's specifically tied to caching in the interop layer, it really makes it sound like native Carbon will not require copyability, but C++ interop will.

I'd suggest revising the "Large elements and Optional" section to explicitly talk about the fact that this design may force `Next` to make a copy, and hence require `ElementType` to be copyable. Then down here you can say that the caching requires `ElementType` to be movable, and note that this won't matter in practice if Carbon requires it to be copyable anyway.

      : container_(other.container_), cursor_(other.cursor_),
      value_(other) {}

  auto operator*() const -> const typename Iterate::ElementType& {
    assert(cursor_);
    return value_;
  }
  auto operator++() -> IterateIterator<Iterate>& {
    next();
    return *this;
  }
  auto operator==(const IterateIterator<Iterate>& rhs) -> bool {
    return container_ == rhs.container_ && cursor_ == rhs.cursor_;
  }
  auto operator!=(const IterateIterator<Iterate>& rhs) -> bool {
    return !(*this == rhs);
  }

 private:
  void next() {
    assert(cursor_);
    if (const auto v = container_->Next(*cursor_); v) {

>       value_ = *v;

Reviewer: ```suggestion
      value_ = *std::move(v);
```

This will enable the code to work with move-only types.

Author: Good point, thanks.
